### PR TITLE
Implement ADC for AVR

### PR DESCRIPTION
**Pull request by deadprogram**
This PR adds the ADC capability to the AVR machine. It also adds a small example program:

```go
package main

import (
	"machine"
	"time"
)

// This example assumes that an analog sensor such as a rotary dial is connected to pin ADC0.
// When the dial is turned past the midway point, the built-in LED will light up.

func main() {
	machine.InitADC()

	led := machine.GPIO{machine.LED}
	led.Configure(machine.GPIOConfig{Mode: machine.GPIO_OUTPUT})

	sensor := machine.ADC{machine.ADC2}
	sensor.Configure()

	for {
		val := sensor.Get()
		if val < 512 {
			led.Low()
		} else {
			led.High()
		}
		time.Sleep(time.Millisecond * 100)
	}
}
```

### CODE REVIEW

File: src/examples/serial/serial.go [hunk 1]
```diff
@@ -0,0 +1,10 @@
+package main
+
+import "time"
+
+func main() {
+	for {
+		println("hello world!")
+		time.Sleep(time.Second)
```
Comment (aykevl) [inline src/examples/serial/serial.go]: I think this example was included by accident? It doesn't seem to belong in this PR (but a separate PR with this would be nice!).

File: src/machine/machine_avr.go [hunk 2]
```diff
@@ -124,3 +124,47 @@ func (pwm PWM) Set(value uint16) {
...
+	return // no pin specific setup on AVR machine.
+}
+
+// Get returns the current value of a ADC pin. The AVR will return a 10bit value ranging
+// from 0-1023.
+func (a ADC) Get() uint16 {
+	// set the analog reference (high two bits of ADMUX) and select the
+	// channel (low 4 bits), masked to only turn on one ADC at a time.
+	// this also sets ADLAR (left-adjust result) to 0 (the default).
+	*avr.ADMUX = avr.RegValue(avr.ADMUX_REFS0 | (a.Pin & 0x07))
```
Comment (aykevl) [inline src/machine/machine_avr.go]: Can you enable ADLAR here, to get a result in the full 16-bit scale (with the low bits zero)?
Different devices vary in ADC precision but most often fall between 8 and 16 bits (nrf has 12 bits, for example) so the API would be most consistent with an ADC that uses the full 16-bit range.
And a small nitpick: the comment has full sentences but no capitals in front of them.

File: src/examples/serial/serial.go (hunk 1 above)
Comment (deadprogram) [reply src/examples/serial/serial.go]: Yes that was an error. I can remove the file if you like and submit a separate PR, or just leave it. UP to you.

File: src/machine/machine_avr.go (hunk 2 above)
Comment (deadprogram) [reply src/machine/machine_avr.go]: I am confused by the request for ADLAR here. In either case the catse will be returned as a unit16 ranging from 0-1023 as it only has 10 bits available. 

The normal result for most ADC is a number from 0 to the max scale. Dividing that by the reference voltage gets the effective voltage for the ADC input. The result will fit into the uint16 so... not sure why the result to change the bit order of the internal registers.

Also regarding the comments: I am of the habit to use full sentences for the GoDoc comments, and smaller, uncapitalized for the case that there are inline comments. Is that a problem?

Comment (aykevl) [reply src/machine/machine_avr.go]: Unless I'm misunderstanding the meaning of ADLAR here, I'm pretty sure when set it will return a value using the full 16-bit range but of course still with 10 bits of precision so the 6 lower bits are always 0. [This forum post](https://www.avrfreaks.net/comment/60470#comment-60470) explains it better than I can. Basically, when set, it will have an equivalent effect as shifting the result in ADCL/ADCH left by 6 bits. This is usually done if you only want to read the 8 high bits of the result (for 8 bits precision), but in this case it is an easy way to shift the result left by 6 without any additional effort.

My intention is the same as with ADC: make sure a program can use it without needing to know the range of the ADC. So for example, an Arduino has a range 0..1023 and nRF has a range of 0..4095. When both are scaled to 0..0xffff a user of the API can simply assume that 0x7fff means 2.5V with an input of 0..5V (for example).

Regarding comments: it's a very minor nit and probably personal style, so it's fine. I also write uncapitalized comments, but tend to use proper punctuation when it is a full sentence or more than one sentence.

File: src/examples/serial/serial.go (hunk 1 above)
Comment (aykevl) [reply src/examples/serial/serial.go]: Yes, if you want please remove it. Otherwise I can remove it when merging (or if you don't mind, send it in as a separate commit).

Offtopic: maybe you've tried to print multiple strings in a source file and found a crash. I'm aware of that, but it really is a bug in the AVR backend of LLVM. However, I plan to do an optimization at some point that should have the side effect of making this happen less often. Printing integers shouldn't result in such a crash.

File: src/machine/machine_avr.go (hunk 2 above)
Comment (deadprogram) [reply src/machine/machine_avr.go]: According to my reading of the AVR docs seems like ADLAR is intended for faster 8-bit conversions where the results are put in only one result register. Perhaps a constant per machine with the max value for ADC would be better to solve the scaling calculation.

Comment (deadprogram) [reply src/machine/machine_avr.go]: After reading that doc it seems unnecessary to  set ADLAR as we will have to read both registers in any case.

File: src/examples/serial/serial.go (hunk 1 above)
Comment (deadprogram) [reply src/examples/serial/serial.go]: I noticed a different problem, that this did not work:

```go
val := uint16(1000)
println("value is", val)
```

The output was:
```
255
```

If I did same but using uint8 it worked as expected.

Comment (aykevl) [reply src/examples/serial/serial.go]: That's a bug, I'll have to look into it.

File: src/machine/machine_avr.go (hunk 2 above)
Comment (aykevl) [reply src/machine/machine_avr.go]: ADLAR is indeed *meant* for faster 8-bit reads. But that doesn't mean it can't be used in a different way, namely to shift the output value (as both ADCH and ADCL are shifted).

Basically I'm arguing two things:
1. The output should be in the range 0..0xffff for easier portability. A possible way to implement that is to shift the result on AVR left by 6 bits, like so:
    ```go
    low := uint16(*avr.ADCL)
    high := uint16(*avr.ADCH)
    return (low <<6) | (high<<14))
    ```
2. The above result can be obtained in a more efficient way by setting the ADLAR bit, avoiding a shift operation.

Yes we can provide constants for the max value, and that would certainly be a good idea. But I think it is easier to write portable programs when all chips use the same 16-bit ADC range (but with only the available precision).

Comment (aykevl) [reply src/machine/machine_avr.go]: > [...] where the results are put in only one result register.

That's not true: both registers are still set. The only difference is which bits are set in the result register. See the tables here: http://www.robotplatform.com/knowledge/ADC/adc_tutorial_3.html

File: src/examples/serial/serial.go (hunk 1 above)
Comment (deadprogram) [reply src/examples/serial/serial.go]: OK removed the serial example. I will submit a separate PR.

File: src/machine/machine_avr.go (hunk 2 above)
Comment (deadprogram) [reply src/machine/machine_avr.go]: Seems like turning on the ADLAR bit and using the shifting code you suggested is not working for me as expected. Perhaps you would like to simply make that change yourself? I am not getting valid numbers like I was expecting.

File: src/examples/serial/serial.go (hunk 1 above)
Comment (deadprogram) [reply src/examples/serial/serial.go]: My example above to repro the `println()` bug was flawed. Now that I am back at computer, here is a way to repro using this branch:

```go
package main

import (
	"machine"
	"time"
)

func main() {
	machine.InitADC()

	led := machine.GPIO{machine.LED}
	led.Configure(machine.GPIOConfig{Mode: machine.GPIO_OUTPUT})

	sensor := machine.ADC{machine.ADC2}
	sensor.Configure()

	for {
		val := sensor.Get()
		if val < 512 {
			led.Low()
		} else {
			led.High()
		}

		println("Value is", val) // this prints only the numeric value of val, not the string prefix "Value is"
		time.Sleep(time.Millisecond * 100)
	}
}
```

File: src/machine/machine_avr.go (hunk 2 above)
Comment (aykevl) [reply src/machine/machine_avr.go]: I'm getting the numbers as expected, so not quite sure where it went wrong with you. When connected to 3.3V I get 0x02a0 and 0x02a1 and when I set ADLAR I get 0xa800 and 0xa840 - exactly the shifted values (`val << 6`). The patch is https://github.com/aykevl/tinygo/commit/58b536619dcf4426de9ef1181877766e8c75615b.

Hint: you can print hexadecimal pointer-sized numbers by printing a number cast to uintptr.

**Comment by aykevl** [comment]
Merged in https://github.com/aykevl/tinygo/commit/40f834d58fee1daa406f0f9edf01e393b3eb7f76.

**Comment by aykevl** [comment]
I accidentally removed the serial example again while merging this PR (oops!). I've added it back in 486ea1b8eae195c6d1354ccf3bc868d75a4954df as I didn't dare to do a force-push.
Lesson learned: don't merge commits after a long tiring day...
